Add cert-manager plugin

yolossn opened the pull request:
# Testing the Cert-Manager Plugin

## Prerequisites
- A Kubernetes cluster with cert-manager installed
  - If you need to install cert-manager, follow the [official installation guide](https://cert-manager.io/docs/installation/)
- Node.js and npm installed on your system

## Steps to Test

1. Clone the plugins repository:
   ```bash
   git clone https://github.com/headlamp-k8s/plugins.git
   ```

2. Switch to the cert-manager branch:
   ```bash
   git checkout cert-manager
   ```

3. Navigate to the cert-manager plugin directory:
   ```bash
   cd cert-manager
   ```

4. Install the required dependencies:
   ```bash
   npm install
   ```

5. Start the plugin in development mode:
   ```bash
   npm run start
   ```

6. Launch Headlamp. You should now see "Cert Manager" in the sidebar.
## Optional: Generate Cert-Manager Types
To test the plugin with sample cert-manager resources:

1. Navigate to the test-files directory:
   ```bash
   cd test-files
   ```

2. Apply the sample configurations to your cluster:
   ```bash
   kubectl apply -f clusterIssuer.yaml
   kubectl apply -f issuer.yaml
   kubectl apply -f certificate.yaml
   kubectl apply -f order.yaml
   kubectl apply -f app.yaml
   ```

This will create:
- A ClusterIssuer for Let's Encrypt staging
- An Issuer for Let's Encrypt staging
- Two Certificate resources (one using ClusterIssuer, one using Issuer)
- An Order resource
- Sample Nginx deployment with Ingress configurations

Demo

https://github.com/user-attachments/assets/3bd40e22-94bc-4624-b7be-d303e3688dd9

juv commented:
This is great 👍

joaquimrocha reviewed (changes requested):
Left a couple of small comments. The plugin should also display a special error if the cert-manager is not installed.

yolossn commented:
<img width="1624" alt="image" src="https://github.com/user-attachments/assets/0fc46f40-fc76-45e1-9646-4ee7a4ffd54d" />

Addressed the comments and added cert-manager not detected banner.

knrt10 reviewed (commented):
Awesome work @yolossn. Apart from some suggestions, everything looks great

yolossn commented on cert-manager/package.json:
```diff
@@ -0,0 +1,37 @@
+    "@kinvolk/headlamp-plugin": "^0.10.0",
+    "@headlamp-k8s/eslint-config": "^0.7.0"
+  }
+}
```
the default project formatting setting is removing the new line.

yolossn commented on cert-manager/README.md:
```diff
@@ -0,0 +1,165 @@
+
+5. **Secret**
+   - Final output containing:
+   - The private key
+   - The signed certificate
+   - The CA certificate chain
+   - Created/updated once the Challenge is successful
```
![image](https://github.com/user-attachments/assets/702506df-9ee3-4c5a-b3a1-3a67d451d8da)
In preview it is appearing properly.

sniok commented on cert-manager/src/components/certificateRequests/List.tsx:
```diff
@@ -0,0 +1,56 @@
+import { NotInstalledBanner } from '../common/NotInstalledBanner';
+
+export function CertificateRequestsList() {
+  const [isManagerInstalled, setIsManagerInstalled] = useState(false);
```
this part (useState + useEffect) is repeated in each `List` component. please extract it into a it's own hook.

this starts with value false which will render NotInstalledBanner by default, even if it is installed. I'd recommend adding additional field `isManagerCheckLoading` that will keep track whether the check is still loading. this will avoid flashing "NotInstalledBanner" 

also this check is loaded every time you enter the page, it should be fetched once and remember the value

yolossn replied:
Created a separate hook for this.

sniok commented on cert-manager/src/index.tsx:
```diff
@@ -0,0 +1,154 @@
+});
+
+// 1. certificates.cert-manager.io
+registerSidebarEntry({
```
this pattern repeats here, `registerSidebarEntry` + `registerRoute` for Details and List could be put in a function

sniok commented on cert-manager/src/types/certificate.ts:
```diff
@@ -0,0 +1,58 @@
+export class Certificate extends KubeObject<CertManagerCertificate> {
+  static kind = 'Certificate';
+  static apiName = 'certificates';
+  static apiVersion = ['cert-manager.io/v1'];
```
apiVersion can be just a string, no need for an array here

sniok commented on cert-manager/src/utils.ts:
```diff
@@ -0,0 +1,12 @@
+import { ApiProxy } from '@kinvolk/headlamp-plugin/lib';
```
this file could be named isCertManagerInstalled since there's nothing else here

sniok commented on cert-manager/src/types/certificate.ts:
```diff
@@ -0,0 +1,58 @@
+import { KubeObject } from '@kinvolk/headlamp-plugin/lib/k8s/cluster';
```
/types/ folder name is a bit misleading. types usually means only typescript type definitions, here we define classes. Maybe /model/ would be better?

yolossn replied:
changed it to resource as in k8s follows that naming convention.

sniok reviewed (changes requested):
good stuff! left a couple of comments

yolossn commented on cert-manager/src/components/certificateRequests/Detail.tsx:
```diff
@@ -0,0 +1,171 @@
+                <Tooltip title={item.spec.request}>{item.spec.request.substring(0, 30)}...</Tooltip>
+                <IconButton
+                  onClick={() => {
+                    navigator.clipboard.writeText(item.spec.request);
```
handled the error case

yolossn commented on cert-manager/src/components/challenges/Detail.tsx:
```diff
@@ -0,0 +1,151 @@
+                <Tooltip title={item.spec.key}>{item.spec.key.substring(0, 30)}...</Tooltip>
+                <IconButton
+                  onClick={() => {
+                    navigator.clipboard.writeText(item.spec.key);
```
done

yolossn commented on cert-manager/src/components/orders/Detail.tsx:
```diff
@@ -0,0 +1,170 @@
+
+                <IconButton
+                  onClick={() => {
+                    navigator.clipboard.writeText(item.spec.request);
```
done

yolossn commented on cert-manager/src/components/certificateRequests/List.tsx:
```diff
@@ -0,0 +1,56 @@
+        {
+          id: 'approved',
+          label: 'Approved',
+          getValue: certificateRequest => certificateRequest.approved,
```
fixed

illume commented on cert-manager/README.md:
````diff
@@ -0,0 +1,165 @@
+
+```
+
+## Developing Headlamp plugins
````
Can you please remove this section?
(It's just there for new plugin authors)

illume commented on cert-manager/README.md:
```diff
@@ -0,0 +1,165 @@
+# cert-manager
+
+cert-manager plugin for Headlamp.
```
Can you please add a bit more here? So people new to cert-manager will have more of an idea. Also what this plugin does for users evaluating the plugin.

- A screenshot or your video would be helpful too.
- What about adding a section for testing this plugin from the PR description?

@vyncent-t I know you have some good ideas on plugin README improvements... can you please review this README?

joaquimrocha replied:
The README has been updated with those suggestions.

illume reviewed (changes requested):
Left a couple of notes on some possible improvements for the README.

Documenting the components/functions/fields would be useful.
